src/app/calculate.service.ts

```diff
@@ -0,0 +1,305 @@
+  private getSocialPercent(year: number, older: boolean, social: boolean): number {
+    /*
+    * JSON properties for socialPercent object
+    * rate: Higher full rate including social contributions to be used to get proportion
+    * social: Percentage of social contributions (AOW + Anw + Wlz)
+    * older: Percentage for retirement age (Anw + Wlz, no contribution to AOW)
+    */
+    let bracket: any = CONSTANTS.socialPercent[year][0],
+      percentage = 1;
+    if (!social) {
+      percentage = (bracket.rate - bracket.social) / bracket.rate; // Removing AOW + Anw + Wlz from total
+    } else if (older) {
+      percentage = (bracket.rate + bracket.older - bracket.social) / bracket.rate; // Removing only AOW from total
+    }
+    return percentage;
+  }
+
+  // Special Rate for One-off Rewards (Loonheffing Bijzonder Tarief)
+  // https://www.belastingdienst.nl/wps/wcm/connect/bldcontentnl/themaoverstijgend/rekenhulpen/loonbelastingtabellen
+  private getSpecialRate(year: number, salary: number, older: boolean, social: boolean): number {
+    let rate = 0;
+    if (social) {
+      rate = this.getRates(CONSTANTS.specialRate[year], salary, (older) ? 'older' : 'rate', false);
+    }
+    return rate;
+  }
+
+  // https://www.belastingdienst.nl/wps/wcm/connect/bldcontentnl/themaoverstijgend/brochures_en_publicaties/handboek-loonheffingen-2017
+  private getRates(brackets: any, salary: number, type: string = 'rate', cumulative: boolean = true): number {
+    let amount: number = 0;
+    let tax: number, delta: number, percent: boolean;
+
+    brackets.some((bracket, index) => {
+      bracket.max = bracket.max || Infinity; // Consider infinity when no upper bound
+      tax = (type && bracket[type]) ? bracket[type] : bracket['rate'];
+      salary = Math.trunc(salary);
+      if (cumulative) {
+        delta = bracket.max - bracket.min;
+        percent = (tax != 0 && tax > -1 && tax < 1); // Check if rate is percentage or fixed
+        if (salary <= delta) {
+          if (percent) {
+            amount += Math.trunc((salary * 100) * tax) / 100; // Round down at 2 decimal places
+          } else {
+            amount = tax;
+          }
+          //console.log(index, salary, delta, tax, percent, amount);
+          return true; // Break loop when reach last bracket
+        } else {
+          if (percent) {
+            amount += (delta * tax);
+          } else {
+            amount = tax;
+          }
+          salary -= delta;
+        }
+      } else if (salary >= bracket.min && salary <= bracket.max) {
+        amount = tax;
+        return true;
+      }
+    });
+    return amount;
+  }
+
+  // Approximation using Newton-Raphson method with numerical derivative
+  // https://github.com/scijs/newton-raphson-method
+  private newtonRaphson(f: any, x0: number, options: any): number {
```

Comment:
you are probably right. i just tried to keep it simple and with less files and folder as possible